_posts/2022-02-17-passwordless-auth-with-express-and-postgres.md

```diff
@@ -0,0 +1,563 @@
+
+### POST /register
+
+Now our first endpoint that actually does something to the database. We have two objectives here:
+
+1. Make sure the user isn't already registered
+2. Register the user
+
+In both cases above we will redirect to the `POST /login` page afterwards to give our users the cleanest login experience (without having to re-enter their email address).
+
+```js
+router.post("/register", async (req, res) => {
+  const { email } = req.body;
+
+  // Check if a user exists
+  const users = await Queries.Users.Get.ByEmail(email);
+
+  // If not, create one
+  if (users.length === 0) {
+  	await Queries.Users.Create(email);
+  }
+
+  // Redirect to /login POST
+  return res.redirect(307, "/users/login");
+});
+```
+
+Another reader exercise exists above -- validate that email address before creating a user.
+
+### POST /login
+
+This is where the one-time password gets generated and emailed to the user. First we need to make sure the appropriate libraries are imported for these actions:
+
+```js
+const bcrypt = require("bcrypt");
+const { SMTPClient } = require("emailjs");
+```
+
+Now that we have the appropriate libraries let's create a function to handle sending the email to the user:
+
+```js
+// Grab the environment variables
+const { EMAIL_USER, EMAIL_PASSWORD, EMAIL_HOST, EMAIL_FROM } = process.env;
+
+// Create an email client using emailjs
+const client = new SMTPClient({
+  user: EMAIL_USER,
+  password: EMAIL_PASSWORD,
+  host: EMAIL_HOST,
+  ssl: true,
+});
+
+// Send the email
+const sendEmail = async (to, subject, data) => {
+  await client.sendAsync({
+    from: EMAIL_FROM,
+    to,
+    subject,
+    attachment: [{ data, alternative: true }],
+  });
+};
+```
+
+The code above should be fairly straight-forward. We create an email client and then build a function that wraps sending emails. Now we just need to pass in the recipient, subject, and HTML data for the email.
+
+The login step is a little more complex, we we'll go through it line-by-line to make sure everything is understood.
+
+First thing is to make sure the user exists in the database:
+
+```js
+router.post("/login", async (req, res) => {
+  const { email } = req.body;
+
+  const users = await Queries.Users.Get.ByEmail(email);
+
+  if (users.length === 0) throw { message: "No user found" };
+
+  // There should only be one result, so grab it
+  const user = users[0];
+});
+```
+
+A more complete app will redirect to the registration screen instead of throwing an error here.
+
+Next we need to generate the one-time password. For this we'll be generating a six character alphanumeric passphrase, then encrypting it for storage in the database.
+
+```js
+router.post("/login", async (req, res) => {
+  ...
+  
+  // Generate OTP
+  const code = Math.random().toString(32).slice(2, 8).toUpperCase();
```

Comment:
Suggestion; this isn't an cryptographically secure way to generate random numbers, it might be worth calling that out or using Bcrypt to generate a random number etc.

Reply:
https://security.stackexchange.com/questions/144445/cryptographically-secure-number-generator-for-node-js

Reply:
Fixed